README.md
# Recent Global Posts Widget

**INACTIVE NOTICE: This plugin is unsupported by WPMUDEV, we've published it here for those technical types who might want to fork and maintain it for their needs.**

Show recent posts from all the sites across an entire network – hugely powerful, massively configurable.

* Set the number of posts displayed 
* Display avatars and preset avatar size 
* Display latest posts from across a network 
* Includes widget for displaying latest posts 
* Set number of title and content characters 
* Display custom post types 

## Recent Global Posts Widget shows all the latest posts from across your entire Multisite or BuddyPress network on your main site – simply, efficiently and quickly.

### Amazing Post Indexer Extension

##### This simple powerful plugin allows you to:

* Use a simple widget to display the **latest posts across your entire network**
* Choice of displaying **title + content, title only or content only**
* Option to change the** number of posts displayed  
**
* Set **number of title and content characters allowed**
* Option to **display avatars** and preset avatar size
* Ability to **display custom post types**

### Familiar Settings

A Recent Global Posts Widget id added to** ‘**Widgets’ in the dashboard of your main site.

![Recent Global Posts Widget - Widget Settings][35]

Drag-n-drop to activate and simple configuration

Drag the Recent Global Posts Widget into your sidebar, choose your configuration options and the latest posts from across your entire network will be pulled into your main site.

![Recent Global Posts Widget In Sidebar][36]

A powerful tool for discovering new content

Recent Global Posts Widget adds a fun way for users to see and discover new content from across your entire network.


### To Install:

1\. Install the Post Indexer

2.  Install the Recent Posts Plugin.

* Once uploaded visit **Network Admin -> Plugins** to Network Activate the Recent Global Posts Widget Plugin.
* Your Recent Global Post widget is added to **Appearance > Widgets **of your main site.
* By default, the widget is available only on the main site of your network. This behavior can be changed by following the instructions below to edit your _wp-config.php_ file.
* The Recent Global posts widget list public posts only.  Posts from privates sites aren’t displayed.

3.  Install the Avatars plugin (if you want to display avatars).

### To Use:

1.  Once installed you just go **Appearance > Widgets.**

2.  Add the Recent Global posts widget to your sidebar.

3\. Check out the configuration options below:

![Recent Global Posts Widget - Widget Settings][35]

Once configured, it could look like this in your sidebar:

![Recent Global Posts Widget In Sidebar][36]

### Enabling Widget for all sites

By default the Recent Global Posts widget is only enabled for use by the man site.

You can enable it for all sites on your network as follows:

1\. Open up the ****wp-config.php** **in the root of your WordPress install

2\. Just before the line that says _That’s all, stop editing!_, add the following:
    
    
    define('RECENT_GLOBAL_POSTS_WIDGET_MAIN_BLOG_ONLY', false);

3\. Save and re-upload your amended wp-config.php file

### Recent Global Posts Widget

#### Quality development and features including

* Display latest posts from across your network on main site 
* Bundled with simple widget to display latest posts 
* Choice of displaying title and content, title only or content only 
* Option to change the number of posts displayed 
* Set number of title and content characters allowed 
* Option to display avatars and preset avatar size 
* Display custom post types 
* Greater control over your user experience 
* Simple configuration and easy to use

[35]: https://premium.wpmudev.org/wp-content/uploads/2009/03/recent-global-posts-widget-3050-widget-settings.png
[36]: https://premium.wpmudev.org/wp-content/uploads/2009/03/recent-global-posts-widget-3050-front.png


changelog.txt
=== Recent Global Comments Widget ===
Authors: Paul Menard (Incsub)

== Changelog ==

= 1.0.4.2 =
* Added support for I18N language files
* Added optional define 'RECENT_GLOBAL_COMMENTS_WIDGET_MAIN_BLOG_ONLY' which can be used to control limiting the widget for display on main site or available on all sites. 
* Test for WordPress 3.6.1

= 1.0.4.1 =
* IMPORTANT security update
* XSS prevention patches
* Update WPMU DEV Update Notification

= 1.0.4 =
* Compatibility update for WordPress 3.1

= 1.0.3 =
* Fixed avatar display issue

= 1.0.2 =
* Fixed avatar link issue

= 1.0.1 =
* Fixed bug that was resulted in spam comments being displayed

= 1.0.0 =
* Initial Release

widget-recent-global-comments.php
<?php
/*
Plugin Name: Recent Comments Widget
Description:
Author: Andrew Billits (Incsub)
Version: 1.0.3
Author URI:
*/

/* 
Copyright 2007-2009 Incsub (http://incsub.com)

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License (Version 2 - GPLv2) as published by
the Free Software Foundation.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

//------------------------------------------------------------------------//
//---Config---------------------------------------------------------------//
//------------------------------------------------------------------------//
$recent_global_comments_widget_main_blog_only = 'yes'; //Either 'yes' or 'no'
//------------------------------------------------------------------------//
//---Hook-----------------------------------------------------------------//
//------------------------------------------------------------------------//

//------------------------------------------------------------------------//
//---Functions------------------------------------------------------------//
//------------------------------------------------------------------------//
function widget_recent_global_comments_init() {
	global $wpdb, $recent_global_comments_widget_main_blog_only;
		
	// Check for the required API functions
	if ( !function_exists('register_sidebar_widget') || !function_exists('register_widget_control') )
		return;

	// This saves options and prints the widget's config form.
	function widget_recent_global_comments_control() {
		global $wpdb;
		$options = $newoptions = get_option('widget_recent_global_comments');
		if ( $_POST['recent-global-comments-submit'] ) {
			$newoptions['recent-global-comments-title'] = $_POST['recent-global-comments-title'];
			$newoptions['recent-global-comments-number'] = $_POST['recent-global-comments-number'];
			$newoptions['recent-global-comments-content-characters'] = $_POST['recent-global-comments-content-characters'];
			$newoptions['recent-global-comments-avatars'] = $_POST['recent-global-comments-avatars'];
			$newoptions['recent-global-comments-avatar-size'] = $_POST['recent-global-comments-avatar-size'];
		}
		if ( $options != $newoptions ) {
			$options = $newoptions;
			update_option('widget_recent_global_comments', $options);
		}
	?>
				<div style="text-align:left">
                
				<label for="recent-global-comments-title" style="line-height:35px;display:block;"><?php _e('Title', 'widgets'); ?>:<br />
                <input class="widefat" id="recent-global-comments-title" name="recent-global-comments-title" value="<?php echo $options['recent-global-comments-title']; ?>" type="text" style="width:95%;">
                </select>
                </label>
				<label for="recent-global-comments-number" style="line-height:35px;display:block;"><?php _e('Number', 'widgets'); ?>:<br />
                <select name="recent-global-comments-number" id="recent-global-comments-number" style="width:95%;">
                <?php
					if ( empty($options['recent-global-comments-number']) ) {
						$options['recent-global-comments-number'] = 10;
					}
					$counter = 0;
					for ( $counter = 1; $counter <= 25; $counter += 1) {
						?>
                        <option value="<?php echo $counter; ?>" <?php if ($options['recent-global-comments-number'] == $counter){ echo 'selected="selected"'; } ?> ><?php echo $counter; ?></option>
                        <?php
					}
                ?>
                </select>
                </label>
				<label for="recent-global-comments-content-characters" style="line-height:35px;display:block;"><?php _e('Content Characters', 'widgets'); ?>:<br />
                <select name="recent-global-comments-content-characters" id="recent-global-comments-content-characters" style="width:95%;">
                <?php
					if ( empty($options['recent-global-comments-content-characters']) ) {
						$options['recent-global-comments-content-characters'] = 50;
					}
					$counter = 0;
					for ( $counter = 1; $counter <= 500; $counter += 1) {
						?>
                        <option value="<?php echo $counter; ?>" <?php if ($options['recent-global-comments-content-characters'] == $counter){ echo 'selected="selected"'; } ?> ><?php echo $counter; ?></option>
                        <?php
					}
                ?>
                </select>
                </label>
				<label for="recent-global-comments-avatars" style="line-height:35px;display:block;"><?php _e('Avatars', 'widgets'); ?>:<br />
                <select name="recent-global-comments-avatars" id="recent-global-comments-avatars" style="width:95%;">
                <option value="show" <?php if ($options['recent-global-comments-avatars'] == 'show'){ echo 'selected="selected"'; } ?> ><?php _e('Show'); ?></option>
                <option value="hide" <?php if ($options['recent-global-comments-avatars'] == 'hide'){ echo 'selected="selected"'; } ?> ><?php _e('Hide'); ?></option>
                </select>
                </label>
				<label for="recent-global-comments-avatar-size" style="line-height:35px;display:block;"><?php _e('Avatar Size', 'widgets'); ?>:<br />
                <select name="recent-global-comments-avatar-size" id="recent-global-comments-avatar-size" style="width:95%;">
                <option value="16" <?php if ($options['recent-global-comments-avatar-size'] == '16'){ echo 'selected="selected"'; } ?> ><?php _e('16px'); ?></option>
                <option value="32" <?php if ($options['recent-global-comments-avatar-size'] == '32'){ echo 'selected="selected"'; } ?> ><?php _e('32px'); ?></option>
                <option value="48" <?php if ($options['recent-global-comments-avatar-size'] == '48'){ echo 'selected="selected"'; } ?> ><?php _e('48px'); ?></option>
                <option value="96" <?php if ($options['recent-global-comments-avatar-size'] == '96'){ echo 'selected="selected"'; } ?> ><?php _e('96px'); ?></option>
                <option value="128" <?php if ($options['recent-global-comments-avatar-size'] == '128'){ echo 'selected="selected"'; } ?> ><?php _e('128px'); ?></option>
                </select>
                </label>
				<input type="hidden" name="recent-global-comments-submit" id="recent-global-comments-submit" value="1" />
				</div>
	<?php
	}
// This prints the widget
	function widget_recent_global_comments($args) {
		global $wpdb, $current_site;
		extract($args);
		$defaults = array('count' => 10, 'username' => 'wordpress');
		$options = (array) get_option('widget_recent_global_comments');

		foreach ( $defaults as $key => $value )
			if ( !isset($options[$key]) )
				$options[$key] = $defaults[$key];

		?>
		<?php echo $before_widget; ?>
			<?php echo $before_title . __($options['recent-global-comments-title']) . $after_title; ?>
            <br />
            <?php
				//=================================================//
				$query = "SELECT * FROM " . $wpdb->base_prefix . "site_comments WHERE blog_public = '1' AND comment_approved = '1' AND comment_type != 'pingback' ORDER BY comment_date_stamp DESC LIMIT " . $options['recent-global-comments-number'];
				$comments = $wpdb->get_results( $query, ARRAY_A );
				if (count($comments) > 0){
					echo '<ul>';
					foreach ($comments as $comment){
						echo '<li>';
						if ( $options['recent-global-comments-avatars'] == 'show' ) {
							if ( !empty( $comment['comment_author_user_id'] ) ) {
								$id_or_email = $comment['comment_author_user_id'];
							} else {
								$id_or_email = $comment['comment_author_email'];
							}
							echo '<a href="' . $comment['comment_post_permalink'] . '">' . get_avatar( $id_or_email, $options['recent-global-comments-avatar-size'], '' ) . '</a>';
							echo ' ';
						}
						echo substr( strip_tags( $comment['comment_content'] ), 0, $options['recent-global-comments-content-characters'] );
						echo ' (<a href="' . $comment['comment_post_permalink'] . '#comment-' . $comment['comment_id'] . '">' . __('More') . '</a>)';
						echo '</li>';
					}
					echo '</ul>';
				}
				//=================================================//
			?>
		<?php echo $after_widget; ?>
<?php
	}
	// Tell Dynamic Sidebar about our new widget and its control
	if ( $recent_global_comments_widget_main_blog_only == 'yes' ) {
		if ( $wpdb->blogid == 1 ) {
			register_sidebar_widget(array(__('Recent Global Comments'), 'widgets'), 'widget_recent_global_comments');
			register_widget_control(array(__('Recent Global Comments'), 'widgets'), 'widget_recent_global_comments_control');
		}
	} else {
		register_sidebar_widget(array(__('Recent Global Comments'), 'widgets'), 'widget_recent_global_comments');
		register_widget_control(array(__('Recent Global Comments'), 'widgets'), 'widget_recent_global_comments_control');
	}
}

add_action('widgets_init', 'widget_recent_global_comments_init');

?>

recent-global-comments-widget.php
<?php
/*
Plugin Name: Recent Global Comments Widget
Plugin URI: http://premium.wpmudev.org/project/recent-global-comments-widget
Description: Display all the latest comments from across your entire WordPress Multisite network - using a simple but powerful widget!
Author: Paul Menard (Incsub)
Author URI: http://ivan.sh
Version: 1.0.4.2
Network: true
WDP ID: 65
Text Domain: recent-global-comments-widget
Domain Path: languages
*/

/* 
Copyright 2007-2011 Incsub (http://incsub.com)

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License (Version 2 - GPLv2) as published by
the Free Software Foundation.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

// Support for WPMU DEV Dashboard plugin
include_once( dirname(__FILE__) . '/lib/dash-notices/wpmudev-dash-notification.php');

function recent_global_comment_widget_init_proc() {
	
	/* Setup the tetdomain for i18n language handling see http://codex.wordpress.org/Function_Reference/load_plugin_textdomain */
    load_plugin_textdomain( 'recent-global-comments-widget', false, dirname( plugin_basename( __FILE__ ) ) . '/languages/' );
}
add_action( 'init', 'recent_global_comment_widget_init_proc' );


//------------------------------------------------------------------------//
//---Config---------------------------------------------------------------//
//------------------------------------------------------------------------//

// Default setting to control user or widget. 'yes' is main blog only. 'no' widget will show for all blogs
$recent_global_comments_widget_main_blog_only = 'yes'; //Either 'yes' or 'no'

// Can also set this in the wp-config.php or theme using 
// define('RECENT_GLOBAL_COMMENTS_WIDGET_MAIN_BLOG_ONLY', 'yes'); // or 'no'
if (defined('RECENT_GLOBAL_COMMENTS_WIDGET_MAIN_BLOG_ONLY'))
	$recent_global_comments_widget_main_blog_only = strtolower(RECENT_GLOBAL_COMMENTS_WIDGET_MAIN_BLOG_ONLY);
		
if ($recent_global_comments_widget_main_blog_only != 'yes')	
	$recent_global_comments_widget_main_blog_only = 'no';

//------------------------------------------------------------------------//
//---Hook-----------------------------------------------------------------//
//------------------------------------------------------------------------//

//------------------------------------------------------------------------//
//---Functions------------------------------------------------------------//
//------------------------------------------------------------------------//

function widget_recent_global_comments_init() {
	global $wpdb, $recent_global_comments_widget_main_blog_only;

	// This saves options and prints the widget's config form.
	function widget_recent_global_comments_control() {
		global $wpdb;
		$options = $newoptions = get_option('widget_recent_global_comments');
		if ( isset( $_POST['recent-global-comments-submit'] ) ) {
			$newoptions['recent-global-comments-title'] 				= sanitize_text_field($_POST['recent-global-comments-title']);
			$newoptions['recent-global-comments-number'] 				= intval($_POST['recent-global-comments-number']);
			$newoptions['recent-global-comments-content-characters'] 	= intval($_POST['recent-global-comments-content-characters']);
			$newoptions['recent-global-comments-avatars'] 				= sanitize_text_field($_POST['recent-global-comments-avatars']);
			$newoptions['recent-global-comments-avatar-size'] 			= intval($_POST['recent-global-comments-avatar-size']);
		}
		if ( $options != $newoptions ) {
			$options = $newoptions;
			update_option('widget_recent_global_comments', $options);
		}
        ?>

        <div style="text-align:left">

            <label for="recent-global-comments-title" style="line-height:35px;display:block;"><?php _e('Title', 'recent-global-comments-widget'); ?>:<br />
            <input class="widefat" id="recent-global-comments-title" name="recent-global-comments-title" value="<?php echo sanitize_text_field($options['recent-global-comments-title']); ?>" type="text" style="width:95%;">
            </select>
            </label>
            <label for="recent-global-comments-number" style="line-height:35px;display:block;"><?php _e('Number', 'recent-global-comments-widget'); ?>:<br />
            <select name="recent-global-comments-number" id="recent-global-comments-number" style="width:95%;">
            <?php
                if ( empty($options['recent-global-comments-number']) ) {
                    $options['recent-global-comments-number'] = 10;
                }
                $counter = 0;
                for ( $counter = 1; $counter <= 25; $counter += 1) {
                    ?>
                    <option value="<?php echo $counter; ?>" <?php if (intval($options['recent-global-comments-number']) == $counter){ echo 'selected="selected"'; } ?> ><?php echo $counter; ?></option>
                    <?php
                }
            ?>
            </select>
            </label>
            <label for="recent-global-comments-content-characters" style="line-height:35px;display:block;"><?php _e('Content Characters', 'recent-global-comments-widget'); ?>:<br />
            <select name="recent-global-comments-content-characters" id="recent-global-comments-content-characters" style="width:95%;">
            <?php
                if ( empty($options['recent-global-comments-content-characters']) ) {
                    $options['recent-global-comments-content-characters'] = 50;
                }
                $counter = 0;
                for ( $counter = 1; $counter <= 500; $counter += 1) {
                    ?>
                    <option value="<?php echo $counter; ?>" <?php if (intval($options['recent-global-comments-content-characters']) == $counter){ echo 'selected="selected"'; } ?> ><?php echo $counter; ?></option>
                    <?php
                }
            ?>
            </select>
            </label>
            <label for="recent-global-comments-avatars" style="line-height:35px;display:block;"><?php _e('Avatars', 'recent-global-comments-widget'); ?>:<br />
            <select name="recent-global-comments-avatars" id="recent-global-comments-avatars" style="width:95%;">
            <option value="show" <?php if ( isset( $options['recent-global-comments-avatars'] ) && $options['recent-global-comments-avatars'] == 'show' ){ echo 'selected="selected"'; } ?> ><?php _e('Show', 'recent-global-comments-widget'); ?></option>
            <option value="hide" <?php if ( isset( $options['recent-global-comments-avatars'] ) && $options['recent-global-comments-avatars'] == 'hide' ){ echo 'selected="selected"'; } ?> ><?php _e('Hide', 'recent-global-comments-widget'); ?></option>
            </select>
            </label>
            <label for="recent-global-comments-avatar-size" style="line-height:35px;display:block;"><?php _e('Avatar Size', 'recent-global-comments-widget'); ?>:<br />
            <select name="recent-global-comments-avatar-size" id="recent-global-comments-avatar-size" style="width:95%;">
            <option value="16" <?php if ( isset( $options['recent-global-comments-avatar-size'] ) && $options['recent-global-comments-avatar-size'] == '16'){ echo 'selected="selected"'; } ?> ><?php _e('16px', 'recent-global-comments-widget'); ?></option>
            <option value="32" <?php if ( isset( $options['recent-global-comments-avatar-size'] ) && $options['recent-global-comments-avatar-size'] == '32'){ echo 'selected="selected"'; } ?> ><?php _e('32px', 'recent-global-comments-widget'); ?></option>
            <option value="48" <?php if ( isset( $options['recent-global-comments-avatar-size'] ) && $options['recent-global-comments-avatar-size'] == '48'){ echo 'selected="selected"'; } ?> ><?php _e('48px', 'recent-global-comments-widget'); ?></option>
            <option value="96" <?php if ( isset( $options['recent-global-comments-avatar-size'] ) && $options['recent-global-comments-avatar-size'] == '96'){ echo 'selected="selected"'; } ?> ><?php _e('96px', 'recent-global-comments-widget'); ?></option>
            <option value="128" <?php if ( isset( $options['recent-global-comments-avatar-size'] ) && $options['recent-global-comments-avatar-size'] == '128'){ echo 'selected="selected"'; } ?> ><?php _e('128px', 'recent-global-comments-widget'); ?></option>
            </select>
            </label>
            <input type="hidden" name="recent-global-comments-submit" id="recent-global-comments-submit" value="1" />
        </div>
        <?php
	}
    
    // This prints the widget
	function widget_recent_global_comments($args) {
		global $wpdb, $current_site;
		extract($args);
		$defaults = array('count' => 10, 'username' => 'wordpress');
		$options = (array) get_option('widget_recent_global_comments');

		foreach ( $defaults as $key => $value ) {
			if ( !isset($options[$key] ) )
				$options[$key] = $defaults[$key];
        }
		?>

		<?php echo $before_widget; ?>
			<?php echo $before_title . $options['recent-global-comments-title'] . $after_title; ?>
            <br />
            <?php
				//=================================================//
				$query = $wpdb->prepare("SELECT * FROM " . $wpdb->base_prefix . "site_comments WHERE blog_public = '1' AND comment_approved = '1' AND comment_type != 'pingback' ORDER BY comment_date_stamp DESC LIMIT %d", $options['recent-global-comments-number']);
				$comments = $wpdb->get_results( $query, ARRAY_A );
				if (count($comments) > 0){
					echo '<ul>';
					foreach ($comments as $comment){
						echo '<li>';
						if ( $options['recent-global-comments-avatars'] == 'show' ) {
							if ( !empty( $comment['comment_author_user_id'] ) ) {
								$id_or_email = $comment['comment_author_user_id'];
							} else {
								$id_or_email = $comment['comment_author_email'];
							}
							echo '<a href="' . $comment['comment_post_permalink'] . '">' . get_avatar( $id_or_email, $options['recent-global-comments-avatar-size'], '' ) . '</a>';
							echo ' ';
						}
						echo substr( strip_tags( $comment['comment_content'] ), 0, $options['recent-global-comments-content-characters'] );
						echo ' (<a href="' . $comment['comment_post_permalink'] . '#comment-' . $comment['comment_id'] . '">' . __('More', 'recent-global-comments-widget') . '</a>)';
						echo '</li>';
					}
					echo '</ul>';
				}
				//=================================================//
			?>
		<?php echo $after_widget; ?>
        <?php
	}

	// Tell Dynamic Sidebar about our new widget and its control
	if ( $recent_global_comments_widget_main_blog_only == 'yes' ) {
		//if ( $wpdb->blogid == 1 ) {
		if ( is_main_site() ) {
			wp_register_sidebar_widget( 'recent_global_comments_widget', __( 'Recent Global Comments', 'recent-global-comments-widget' ), 'widget_recent_global_comments' );
			wp_register_widget_control( 'recent_global_comments_widget', __( 'Recent Global Comments', 'recent-global-comments-widget' ), 'widget_recent_global_comments_control' );
		}
	} else {
        wp_register_sidebar_widget( 'recent_global_comments_widget', __( 'Recent Global Comments', 'recent-global-comments-widget' ), 'widget_recent_global_comments' );
		wp_register_widget_control( 'recent_global_comments_widget', __( 'Recent Global Comments', 'recent-global-comments-widget' ), 'widget_recent_global_comments_control' );
	}
}

add_action('widgets_init', 'widget_recent_global_comments_init');
